tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -11299,6 +11299,36 @@ bool Sema::DiagnoseHLSLMethodCall(const CXXMethodDecl *MD, SourceLocation Loc) {
   return false;
 }
 
+void DiagnoseEntryAttrAllowedOnStage(clang::Sema *self,
+                                     FunctionDecl *entryPointDecl,
+                                     HLSLShaderAttr *shaderAttr) {
+
+  std::string stageName = shaderAttr->getStage();
+
+  // whether or not this attribute should exist on this shader stage
+  bool isValid = false;
+  if (entryPointDecl->hasAttrs()) {
+    for (Attr *pAttr : entryPointDecl->getAttrs()) {
+      switch (pAttr->getKind()) {
+
+      case clang::attr::HLSLWaveSize: {
+        bool isValid = llvm::StringSwitch<bool>(stageName)
```

Comment:
Can't we pass down the ShaderKind and switch on that here instead of using a StringSwitch for each attribute here?